### PR TITLE
[psbt saga 5/4]: Anchor multiple virtual transactions

### PR DESCRIPTION
**Pull request by guggero**
Depends on https://github.com/lightninglabs/taproot-assets/pull/796.

Fixes https://github.com/lightninglabs/taproot-assets/issues/754.
Fixes https://github.com/lightninglabs/taproot-assets/issues/670.

- - -
This change is [Reviewable](https://reviewable.io/reviews/lightninglabs/taproot-assets/821)

### CODE REVIEW

**Review by Roasbeef — APPROVED** [review]
:lgtm: 🏄🏻‍♀️

> Reviewed 2 of 2 files at r1, 2 of 2 files at r2, 2 of 2 files at r3, 1 of 1 files at r4, 6 of 6 files at r5, 2 of 2 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/lightninglabs/taproot-assets/821#-:-NsBekWeAVzuaGwEXbeM:b-zhhdga)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @guggero)

___
*[`asset/asset.go` line 1483 at r1](https://reviewable.io/reviews/lightninglabs/taproot-assets/821#-NsBcMXa2BbpxBGFhWtd:-NsBcMXa2BbpxBGFhWte:b-exkrko) ([raw file](https://github.com/lightninglabs/taproot-assets/blob/4d19d33836635acf66ca9f4d4889563e3e357615/asset/asset.go#L1483)):*
> ```Go
> // The first witness effectively corresponds to the asset's direct lineage.
> func (a *Asset) PrimaryPrevID() (*PrevID, error) {
> 	prevWitnesses := a.Witnesses()
> ```

Nice simplification!

File: itest/multi_send_test.go [hunk 1]
```diff
@@ -0,0 +1,289 @@
...
+		assetXTranche3ID, assetXTranche3.Amount, scriptKey1, 0,
+		anchorInternalKeyDesc1, asset.V0, chainParams,
+	)
+	fundRespSecondBatch := fundPacket(t, aliceTapd, vPktSecondBatch)
+	tranche3, err := tappsbt.Decode(fundRespSecondBatch.FundedPsbt)
+	require.NoError(t.t, err)
+
+	activePackets = append(activePackets, tranche3)
+
+	// We now need to update the outputs of the passive packets to also go
```
Comment (GeorgeTsagk) [inline itest/multi_send_test.go:203]: aren't passive assets supposed to stay "passive" i.e not change?
if the passive assets are going to a different script key, aren't they effectively "active"?

maybe I just can't grasp the phrasing

File: itest/multi_send_test.go [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+
+	// Next, we make sure we can still send out the passive assets.
+	sendAssetAndAssert(
+		ctxt, t, t.tapd, secondTapd, assetP.Amount, 0,
+		assetP.AssetGenesis, assetP, 1, 2, 1,
+	)
+	sendAssetAndAssert(
+		ctxt, t, t.tapd, secondTapd, assetQ.Amount, 0,
+		assetQ.AssetGenesis, assetQ, 2, 3, 2,
+	)
```
Comment (GeorgeTsagk) [inline itest/multi_send_test.go:288]: should we also be doing some checks for the assetY? it's not clear why we minted it together with assetX in the same batches or what its purpose was

**Review by GeorgeTsagk — COMMENTED** [review]
few Qs, also linter seems to be complaining

File: tapfreighter/chain_porter.go [hunk 3]
```diff
@@ -408,13 +408,22 @@ func (p *ChainPorter) storeProofs(sendPkg *sendPackage) error {
 				idx, err)
 		}
 
-		inputs := fn.Map(
-			parcel.Inputs, func(in TransferInput) asset.PrevID {
-				return in.PrevID
-			},
-		)
+		var inputsForAsset []asset.PrevID
```
[the remark this answers is not in the document]
Comment (guggero) [reply tapfreighter/chain_porter.go:411]: Yes, we would add more proofs to the `AdditionalProofs` field of the proof than necessary, leading to an error in the state transition validation within the VM.

File: rpcserver.go [hunk 4]
```diff
@@ -1905,39 +1902,53 @@ func (r *rpcServer) AnchorVirtualPsbts(ctx context.Context,
...
+					"commitment: %w", err)
+			}
 
-		inputCommitments[prevID] = inputCommitment.Commitment
+			inputCommitments[prevID] = inputCommitment.Commitment
+		}
 	}
 
+	rpcsLog.Debugf("Selected %d input commitments for send of %d virtual "+
+		"transactions", len(inputCommitments), len(vPackets))
```
[the remark this answers is not in the document]
Comment (guggero) [reply rpcserver.go:1948]: Good idea, added the additional log as `TRC`.

File: itest/multi_send_test.go [hunk 5]
```diff
@@ -0,0 +1,289 @@
...
+	assetQReq = &mintrpc.MintAsset{
+		AssetType: taprpc.AssetType_NORMAL,
+		Name:      "itest-asset-Q",
+		AssetMeta: dummyMetaData,
+		Amount:    5000,
+	}
+	chainParams = &address.RegressionNetTap
+)
+
+func testAnchorMultipleVirtualTransactions(t *harnessTest) {
```
[the remark this answers is not in the document]
Comment (guggero) [reply itest/multi_send_test.go:85]: Oops, yeah, forgot the godoc. Fixed.

File: itest/multi_send_test.go [hunk 6]
```diff
@@ -0,0 +1,289 @@
...
+	ctxb := context.Background()
+	ctxt, cancel := context.WithTimeout(ctxb, defaultWaitTimeout)
+	defer cancel()
+
+	// In our first batch we create multiple units of the grouped asset X
```
[the remark this answers is not in the document]
Comment (guggero) [reply itest/multi_send_test.go:90]: Yes, it's set to 80. Is your editor configured to use 8 spaces per tab character?

File: itest/multi_send_test.go (hunk 1 above)
Comment (guggero) [reply itest/multi_send_test.go:203]: You're correct, if they go to a different script key, they become active. So in this test we're promoting some of the passive assets to active, which should in the end be reflected in the transfer output table.
Added some comments to make it more clear.

File: itest/multi_send_test.go (hunk 2 above)
Comment (guggero) [reply itest/multi_send_test.go:302]: Correct as well. I initially wanted to add another test case with tombstone outputs, and those two assets (P and Q) were just added for that. So I added that additional case now.

**Review by GeorgeTsagk — APPROVED** [review]
Lgtm :dart:

**Pull request merged**